src/processor.rs

```diff
@@ -210,9 +226,109 @@ fn process_distribute_rewards(
 /// [InitializeHolderRewards](enum.PaladinRewardsInstruction.html)
 /// instruction.
 fn process_initialize_holder_rewards(
-    _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
 ) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let holder_rewards_pool_info = next_account_info(accounts_iter)?;
+    let holder_rewards_info = next_account_info(accounts_iter)?;
+    let token_account_info = next_account_info(accounts_iter)?;
+    let mint_info = next_account_info(accounts_iter)?;
+    let _system_program = next_account_info(accounts_iter)?;
+
+    let token_supply = {
+        let mint_data = mint_info.try_borrow_data()?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+        mint.base.supply
+    };
+
+    let token_account_balance = {
+        let token_account_data = token_account_info.try_borrow_data()?;
+        let token_account = StateWithExtensions::<Account>::unpack(&token_account_data)?;
+
+        // Ensure the provided token account is for the mint.
+        if !token_account.base.mint.eq(mint_info.key) {
+            return Err(PaladinRewardsError::TokenAccountMintMismatch.into());
+        }
+
+        token_account.base.amount
+    };
```

Comment:
> Mathematically, we could say that their unharvested rewards are actually their proportion of the _lamports in the reward account_,

Lamports was what I was going with, yeah. I never even considered using SOL denomination in the program.

> For example, if I have 1 million tokens, and you have 2 million tokens, and there are 9 SOL in total rewards, my last seen is 3 SOL, and yours is 9 SOL, which means you've harvested 6 SOL (2/3 of 9), and I've harvested 1 SOL (1 / 3 of total rewards * 1 / 3 of total supply), so there's only 2 SOL in the account. If all of a sudden we mint 9 million tokens into my account, and I go to claim, the math will say that I can get 10 / 12 of 6 SOL, which is more than the 2 SOL actually left in the account!

I'm glad you brought this up, because this exact situation led me to actually implement the `calculate_reward_share` for all three processor functions: initialize holder rewards, harvest rewards, and transfer hook.

We could omit the calculation from harvest rewards, but then we run into a problem if tokens are ever burnt. Suddenly you have too little `unharvested_rewards` for the mint supply. This might be okay, since you aren't depleting the rewards account, but how does your unharvested amount get adjusted for the new supply without transferring?

It's looking like we'll have to do this calculation in all three places, so that leads me to wonder: why do it here at all? In fact, why even include a "holder rewards" account at all? You could simply just run the calculation on harvest each time. The account _does_ give you a queryable estimate for unharvested rewards, which ideally _shouldn't_ vary too much, but ultimately proper handling of the mint will be key to this system.

The mint supply should be adjusted very carefully - including minting new tokens and burning - so as to not skew holder rewards in either direction.

Reply:
If tokens are burned after your `unharvested_rewards` was calculated, anyone with unharvested rewards is technically getting less than they should, but that's a fine tradeoff, and similar to the situation of minting new tokens to someone else. `unharvested_rewards` shouldn't get adjusted, and it would be dangerous to adjust it. It just means that a harvest might later overdraw the rewards account, in which case we give them whatever's in the reward account but set their last seen to the current amount.

And the system definitely needs a holder rewards account to figure out how much someone has already harvested, and what they're entitled to. The transfer hook ensures that someone doesn't double dip, otherwise the whole system is screwed :sweat_smile:

Reply:
I guess what I meant was "why even use `unharvested_rewards`". The last seen would still be required.

Reply:
Correct, they're both required during harvest. The `unharvested_rewards` is only updated during transfer. The other option is to withdraw rewards during all transfers, but that would require a write lock on the rewards account and effectively force transfers to operate in serial. We could change the design to withdraw rewards on transfer, but I'd prefer to avoid the write lock.

Reply:
> To summarize, I'm thinking we have the unharvested rewards initialized to a proportion of the _lamports_ in the rewards account, and then allow for a situation where the rewards account doesn't have enough SOL to pay the person out. Naturally, the best option is to pre-mint everything before any rewards are ever distributed, but we need to make sure we don't do something too bad here. What do you think?

It sounds like, in general, this is what the current implementation supports, right? The `total_rewards` field in the pool is really just the excess lamports above rent-exemption, and when depositors "distribute" rewards that number increments. Holders track their share of total supply as a share of unharvested rewards, but there's no consideration for minting new tokens.

I think the orchestrators of this system should take special care around managing the token supply, and should be aware of the consequences of supply management decisions. If they decide there should be additional functionality in the protocol to support some supply transitions, it can be added.

All that being said, is there anything left on this change to adjust before merging?

Reply:
The `total_rewards` field is a running counter of *all* rewards that have been distributed, and not `account.lamports - rent_exemption`. If it were to go down, it would be impossible to calculate rewards of individual accounts.

So the initial calculation needs to be adjusted to start unharvested rewards at 0 or a proportion of actual unharvested rewards in the account (ie `account.lamports - rent_exemption`). Setting it to 0 is simpler, but I think the latter is more fair in the case of minting new tokens.

Reply:
Okay, I've added a commit for this and updated the tests to more closely test this calculation.